// esri-loader/dist/esri-loader.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
	typeof define === 'function' && define.amd ? define(['exports'], factory) :
	(factory((global.esriLoader = global.esriLoader || {})));
}(this, (function (exports) { 'use strict';

/*
  Copyright 2017 Esri
  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/
// get the script injected by this library
function getScript() {
    return document.querySelector('script[data-esri-loader]');
}
// has ArcGIS API been loaded on the page yet?
function isLoaded() {
    // TODO: instead of checking that require is defined, should this check if it is a function?
    return typeof window['require'] !== 'undefined' && getScript();
}
// load the ArcGIS API on the page
function bootstrap(callback, options) {
    if (options === void 0) { options = {}; }
    // default options
    if (!options.url) {
        options.url = 'https://js.arcgis.com/4.5/';
    }
    // don't reload API if it is already loaded or in the process of loading
    if (getScript()) {
        if (callback) {
            callback(new Error('The ArcGIS API for JavaScript is already loaded.'));
        }
        return;
    }
    if (options.dojoConfig) {
        // set dojo configuration parameters before loading the script
        window['dojoConfig'] = options.dojoConfig;
    }
    // create a script object whose source points to the API
    var script = document.createElement('script');
    script.type = 'text/javascript';
    script.src = options.url;
    script.dataset['esriLoader'] = 'loading';
    // once the script is loaded...
    script.onload = function () {
        // update the status of the script
        script.dataset['esriLoader'] = 'loaded';
        // we can now use Dojo's require() to load esri and dojo AMD modules
        var _dojoRequire = window['require'];
        if (callback) {
            // let the caller know that the API has been successfully loaded
            // and as a convenience, return the require function
            // in case they want to use it directly
            callback(null, _dojoRequire);
        }
    };
    // handle any script loading errors
    var onScriptError = function (e) {
        if (callback) {
            // pass the error to the callback
            callback(e.error || new Error("There was an error attempting to load " + script.src));
        }
        // remove this event listener
        script.removeEventListener('error', onScriptError, false);
    };
    script.addEventListener('error', onScriptError, false);
    // load the script
    document.body.appendChild(script);
}
function dojoRequire(modules, callback) {
    if (isLoaded()) {
        // already loaded, just call require
        window['require'](modules, callback);
    }
    else {
        // wait for script to load then call require
        var script_1 = getScript();
        if (script_1) {
            // Not yet loaded but script is in the body - use callback once loaded
            var onScriptLoad_1 = function () {
                window['require'](modules, callback);
                script_1.removeEventListener('load', onScriptLoad_1, false);
            };
            script_1.addEventListener('load', onScriptLoad_1, false);
        }
        else {
            // Not bootstrapped
            throw new Error('The ArcGIS API for JavaScript has not been loaded. You must first call esriLoader.bootstrap()');
        }
    }
}
// export a namespace to expose all functions
var esriLoader = {
    isLoaded: isLoaded,
    bootstrap: bootstrap,
    dojoRequire: dojoRequire
};

exports.isLoaded = isLoaded;
exports.bootstrap = bootstrap;
exports.dojoRequire = dojoRequire;
exports['default'] = esriLoader;

Object.defineProperty(exports, '__esModule', { value: true });

})));
