// angular-esri-components/src/lib/widgets/layers-toggle/layers-toggle.component.js
import { Component, Input } from '@angular/core';
import { EsriMapService } from './../../core/index';
var LayersToggleComponent = /** @class */ (function () {
    function LayersToggleComponent(mapService) {
        this.mapService = mapService;
    }
    LayersToggleComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.mapService.isLoaded.subscribe(function () {
            _this.map = _this.mapService.map;
            _this.mapView = _this.mapService.mapView;
        });
    };
    LayersToggleComponent.prototype.onCheck = function ($event, layer) {
        layer.visible = $event.target.checked;
    };
    LayersToggleComponent.prototype.onZoomLayer = function (layer) {
        this.mapView.goTo(layer.fullExtent);
    };
    LayersToggleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'esri-layersToggle',
                    template: "\n    <div esri-customWidget [position]=\"position\">\n      <!-- Only display if there is a map and it has layers -->\n      <div *ngIf=\"map?.layers.length > 0\" class=\"esri-layers-toggle\">\n        <div *ngFor=\"let layer of map.layers.toArray()\">\n          <input type=\"checkbox\" [attr.checked]=\"layer.visible\" (click)=\"onCheck($event, layer)\" /> {{layer.title}}\n          <a href=\"javascript:void(0)\" (click)=\"onZoomLayer(layer)\">Zoom</a>\n        </div>\n      </div>\n    </div>\n  ",
                    styles: ["\n    .esri-layers-toggle {\n      z-index: 99;\n      background-color: white;\n      border-radius: 8px;\n      padding: 10px;\n      opacity: 0.75;\n    }\n  "]
                },] },
    ];
    /** @nocollapse */
    LayersToggleComponent.ctorParameters = function () { return [
        { type: EsriMapService, },
    ]; };
    LayersToggleComponent.propDecorators = {
        "position": [{ type: Input },],
    };
    return LayersToggleComponent;
}());
export { LayersToggleComponent };
