// angular-esri-components/src/lib/esri-map/esri-map.component.js
import { Component, Input, Output, ElementRef, ViewChild, EventEmitter } from '@angular/core';
import { EsriMapService } from './../core/index';
import { environment } from '../../environments/environment';
var EsriMapComponent = /** @class */ (function () {
    function EsriMapComponent(mapService) {
        this.mapService = mapService;
        this.mapViewProperties = {};
        this.mapInit = new EventEmitter();
    }
    EsriMapComponent.prototype.ngOnInit = function () {
        if (this.map) {
            // map is already initialized
            return;
        }
        this.loadMap();
    };
    EsriMapComponent.prototype.loadMap = function () {
        var _this = this;
        var mapPromise;
        // determine if loading a WebMap or creating a custom map
        if (this.mapProperties) {
            mapPromise = this.mapService.loadMap(this.mapProperties, this.mapViewProperties, this.mapEl);
        }
        else if (this.webMapProperties) {
            mapPromise = this.mapService.loadWebMap(this.webMapProperties, this.mapViewProperties, this.mapEl);
        }
        else {
            console.error('Proper map properties were not provided');
            return;
        }
        mapPromise.then(function (mapInfo) {
            _this.map = mapInfo.map;
            _this.mapView = mapInfo.mapView;
            // emit event informing application that the map has been loaded
            // emit event informing application that the map has been loaded
            _this.mapInit.emit({
                map: _this.map,
                mapView: _this.mapView
            });
            _this.mapInit.complete();
        });
    };
    EsriMapComponent.decorators = [
        { type: Component, args: [{
                    selector: 'esri-map',
                    template: "\n    <div class=\"map\" id=\"esri-map\" #map></div>\n  ",
                    styles: ["\n  /* Required CSS for the ArcGIS API for JavaScript */\n  @import '" + environment.arcGisCssUrl + "';\n\n  .map {\n    height: 100%;\n    width: 100%;\n    margin: 0;\n    padding: 0;\n  }\n    "]
                },] },
    ];
    /** @nocollapse */
    EsriMapComponent.ctorParameters = function () { return [
        { type: EsriMapService, },
    ]; };
    EsriMapComponent.propDecorators = {
        "mapEl": [{ type: ViewChild, args: ['map',] },],
        "mapProperties": [{ type: Input },],
        "webMapProperties": [{ type: Input },],
        "mapViewProperties": [{ type: Input },],
        "mapInit": [{ type: Output },],
    };
    return EsriMapComponent;
}());
export { EsriMapComponent };
